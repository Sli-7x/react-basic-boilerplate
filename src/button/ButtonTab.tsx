import * as React from 'react';
import styled from 'styled-components';

const Content = styled.div`
  display: flex;
`;

const StyledTab = styled.div`
  color: ${(props: { selected: boolean }) => (props.selected ? '#fff' : '#282828')};
  background: ${(props: { selected: boolean }) => (props.selected ? '#308c91' : '#fff')};
  padding: ${(props: { selected: boolean }) => (props.selected ? '8px 20px' : '7px 20px')};
  border-right: 1px solid #ccc;
  border-left: 1px solid #ccc;
  cursor: pointer;
  flex: 1;
  text-align: center;
  border: ${(props: { selected: boolean }) => (props.selected ? 'none' : '1px solid #ccc')};
  align-self: center;

  &:first-child {
    border-radius: 5px 0 0 5px;
  }
  &:last-child {
    border-radius: 0 5px 5px 0;
  }
`;

export interface IButtonTabList {
  id: string | number;
  title: string | number;
}

export interface IButtonTabProps {
  onClick: (value: number | string, name: string) => void;
  value: string | number;
  name: string;
  list: IButtonTabList[];
}

export interface IButtonTabProps {}

export class ButtonTab extends React.Component<IButtonTabProps, any> {
  shouldComponentUpdate(nextProps: IButtonTabProps) {
    if (nextProps.value === this.props.value) {
      return false;
    }

    return true;
  }

  handleClick(value: number | string) {
    this.props.onClick(value, this.props.name);
  }

  render() {
    const { value, list } = this.props;

    return (
      <Content>
        {list.map((val: any, i: number) => (
          <StyledTab onClick={this.handleClick.bind(this, val.id)} key={i} selected={value === val.id}>
            {val.title}
          </StyledTab>
        ))}
      </Content>
    );
  }
}
