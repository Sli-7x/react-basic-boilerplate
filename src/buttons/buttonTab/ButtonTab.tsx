import * as React from 'react';
import { Content, StyledTab } from './ButtonTab.style';

export interface IButtonTabList {
  id: string | number;
  title: string | number;
}

export interface IButtonTabProps {
  onClick: (value: number | string, name: string) => void;
  value: string | number;
  name: string;
  list: IButtonTabList[];
}

export interface IButtonTabProps {}

export class ButtonTab extends React.Component<IButtonTabProps, any> {
  shouldComponentUpdate(nextProps: IButtonTabProps) {
    if (nextProps.value === this.props.value) {
      return false;
    }

    return true;
  }

  handleClick(value: number | string) {
    this.props.onClick(value, this.props.name);
  }

  render() {
    const { value, list } = this.props;

    return (
      <Content>
        {list.map((val: any, i: number) => (
          <StyledTab onClick={this.handleClick.bind(this, val.id)} key={i} selected={value === val.id}>
            {val.title}
          </StyledTab>
        ))}
      </Content>
    );
  }
}
