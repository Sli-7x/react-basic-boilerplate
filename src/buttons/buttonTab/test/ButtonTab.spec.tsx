import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { ButtonTab, IButtonTabProps } from '../ButtonTab';

describe('ButtonTab', () => {
  const onClick = jest.fn();

  const props: IButtonTabProps = {
    onClick,
    value: '',
    name: '',
    list: [
      {
        id: 1,
        title: 'test',
      },
    ],
  };

  it('should render correctly', () => {
    const tree = renderer.create(<ButtonTab {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call onClick property', () => {
    const wrapper = shallow(<ButtonTab {...props} />);
    const instance = wrapper.instance() as ButtonTab;
    instance.handleClick('test');
    expect(onClick).toHaveBeenCalled();
  });

  describe('shouldComponentUpdate()', () => {
    it('should not update component', () => {
      const wrapper = mount(<ButtonTab {...props} />);
      const instance = wrapper.instance() as ButtonTab;

      const shouldUpdate = instance.shouldComponentUpdate(props);
      expect(shouldUpdate).toBe(false);
    });

    it('should update component', () => {
      const wrapper = mount(<ButtonTab {...props} />);
      const instance = wrapper.instance() as ButtonTab;

      const shouldUpdate = instance.shouldComponentUpdate({ ...props, value: 'other' });
      expect(shouldUpdate).toBe(true);
    });
  });
});
