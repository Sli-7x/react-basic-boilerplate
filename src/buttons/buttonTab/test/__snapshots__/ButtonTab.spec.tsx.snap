// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonTab should render correctly 1`] = `
<div
  className="sc-bdVaJa kZAgjJ"
>
  <div
    className="sc-bwzfXH gfKqew"
    onClick={[Function]}
    selected={false}
  >
    test
  </div>
</div>
`;
